clang/include/clang/3C/RewriteUtils.h

```diff
@@ -118,11 +118,21 @@ class FunctionDeclReplacement
 
     // If rewriting Parameters, stop at the right parenthesis of the parameters.
     // Otherwise, stop after the return type.
-    // Note: getFunctionDeclarationEnd is used instead of getRParenLoc so that
-    // itypes are deleted correctly when --remove-itypes is used.
-    SourceLocation End = RewriteParams
-                             ? getFunctionDeclarationEnd(Decl, SM)
-                             : Decl->getReturnTypeSourceRange().getEnd();
+    SourceLocation End;
+    if (RewriteParams) {
+      if (auto *BoundsE = Decl->getBoundsExpr())
+        End = BoundsE->getEndLoc();
+      else if (auto *InteropE = Decl->getInteropTypeExpr())
+        End = InteropE->getEndLoc();
+      else
+        End = TypeLoc.getRParenLoc();
+      // SourceLocations are weird and turn up invalid for reasons I don't
+      // understand. Fallback to the original FunctionDecl end function.
+      if (!End.isValid())
+        End = getFunctionDeclarationEnd(Decl, SM);
```

Comment:
Is `getFunctionDeclarationEnd` used elsewhere? It's unfortunate that this condition needs to be here, since it was the offending code, but if it's a hack I think it's better inline rather than exposed to be reused.